src/software/thunderscope/gl/layers/gl_draw_polygon_obstacle.py

```diff
@@ -0,0 +1,145 @@
+from software.thunderscope.binary_context_managers.full_system import ProtoUnixIO
+from software.thunderscope.gl.graphics.gl_polygon import GLPolygon
+from pyqtgraph.Qt import QtGui
+from pyqtgraph.Qt.QtCore import Qt
+from pyqtgraph.opengl import *
+
+from proto.import_all_protos import *
+from software.py_constants import *
+from software.thunderscope.gl.helpers.observable_list import ObservableList
+from software.thunderscope.proto_unix_io import ProtoUnixIO
+
+from software.thunderscope.gl.layers.gl_layer import GLLayer
+from software.thunderscope.gl.helpers.extended_gl_view_widget import MouseInSceneEvent
+
+
+class GLDrawPolygonObstacleLayer(GLLayer):
+    """A layer used to draw polygons that are going to represent obstacles for the trajectroy planner
+    to avoid.
+    """
+
+    def __init__(self, name: str, friendly_io: ProtoUnixIO) -> None:
+        """Initialize this layer
+
+        :param name:            the name of this layer
+        :param friendly_io:     the friendly_io
+        """
+        super().__init__(name)
+
+        self.friendly_io: ProtoUnixIO = friendly_io
+
+        self.current_polygon: GLPolygon = GLPolygon(parent_item=self, line_width=2)
+        # Tuple[float, float] represents a point (x,y)
+        self.points: List[Tuple[float, float]] = []
+
+        self.obstacles: List[Obstacle] = []
+
+        # used for keeping track and rendering multiple polygons
+        self.rendering_polygons: ObservableList = ObservableList(self._graphics_changed)
+
+    def keyPressEvent(self, event: QtGui.QKeyEvent) -> None:
+        """Responding to key events that are going to push obstacles to the stack or add point
+
+        :param event: a qt key event indicating the button that are pressed
+        """
+        if not self.visible():
+            return
+
+        if event.key() == Qt.Key.Key_P:
+            self.push_polygon_to_list()
```

Comment:
tiny nit: i personally like double-clicking as a way to finish a polygon, rather than a key press